forc-plugins/forc-client/src/util/pkg.rs

```diff
@@ -85,3 +102,366 @@ pub(crate) fn built_pkgs(path: &Path, build_opts: &BuildOpts) -> Result<Vec<Arc<
+            i,
+            chunk_contract_ids[i - 1]
+        ));
+    }
+    contract.push_str("}\n\n");
+
+    // ABI
+    contract.push_str("abi LoaderContract {\n");
+    for function in &abi.functions {
+        let inputs = function
+            .inputs
+            .iter()
+            .map(|input| {
+                format!(
+                    "{}: {}",
+                    input.name,
+                    type_id_to_string(input.type_id, &types)
+                )
+            })
+            .collect::<Vec<_>>()
+            .join(", ");
+        let output = type_id_to_string(function.output.type_id, &types);
+        contract.push_str(&format!(
+            "    fn {}({}) -> {};\n",
+            function.name, inputs, output
+        ));
+    }
+    contract.push_str("}\n\n");
+
+    // Implementation
+    contract.push_str("impl LoaderContract for Contract {\n");
+    for function in &abi.functions {
+        let inputs = function
+            .inputs
+            .iter()
+            .map(|input| {
+                format!(
+                    "{}: {}",
+                    input.name,
+                    type_id_to_string(input.type_id, &types)
+                )
+            })
+            .collect::<Vec<_>>()
+            .join(", ");
+        let output = type_id_to_string(function.output.type_id, &types);
+        contract.push_str(&format!(
+            "    fn {}({}) -> {} {{\n",
+            function.name, inputs, output
+        ));
+        contract.push_str(&format!("        run_external{}(", num_chunks));
+        for i in 1..=num_chunks {
+            if i > 1 {
+                contract.push_str(", ");
+            }
+            contract.push_str(&format!("TARGET_{}", i));
+        }
+        contract.push_str(")\n    }\n");
+    }
+    contract.push_str("}\n\n");
+
+    // run_external function
+    contract.push_str(&generate_run_external(num_chunks));
+
+    contract
+}
+
+/// Creates a map from type id to the name of the type, so that we can do ,ook
+fn create_type_map(abi: &ProgramABI) -> HashMap<usize, String> {
+    abi.types
+        .iter()
+        .map(|t| (t.type_id, t.type_field.clone()))
+        .collect()
+}
+
+fn type_id_to_string(type_id: usize, types: &HashMap<usize, String>) -> String {
+    types
+        .get(&type_id)
+        .cloned()
+        .unwrap_or_else(|| format!("Type{}", type_id))
+}
+
+/// Generates an run external function given the `num_targets`. This function
+/// is later used in the generated loader contract.
+fn generate_run_external(num_targets: usize) -> String {
+    let mut func = String::new();
+
+    func.push_str(&format!("fn run_external{}(", num_targets));
+    for i in 1..=num_targets {
+        if i > 1 {
+            func.push_str(", ");
+        }
+        func.push_str(&format!("load_target{}: ContractId", i));
+    }
+    func.push_str(") -> ! {\n");
+
+    // Generate assembly
+    func.push_str("    asm(\n");
```

Comment:
Could you please add a lot more detailed comments about what this assembly code is doing? It might also be nice to add the comment in the generated sway code.